Mathlib/RingTheory/PowerSeries/WellKnown.lean

```diff
@@ -66,6 +67,77 @@ theorem map_invUnitsSub (f : R →+* S) (u : Rˣ) :
 
 end Ring
 
+section invOneSubPow
+
+variable {S : Type*} [CommRing S] (d : ℕ)
+
+/--
+(1 + X + X^2 + ...) * (1 - X) = 1.
+
+Note that the power series `1 + X + X^2 + ...` is written as `mk 1` where `1` is the constant
+function so that `mk 1` is the power series with all coefficients equal to one.
+-/
+theorem mk_one_mul_one_sub_eq_one : (mk 1 : S⟦X⟧) * (1 - X) = 1 := by
+  rw [mul_comm, ext_iff]
+  exact fun n => by
+    by_cases hn : n = 0
+    · subst hn
+      simp only [coeff_zero_eq_constantCoeff, map_mul, map_sub, map_one, constantCoeff_X, sub_zero,
+        one_mul, coeff_one, ↓reduceIte]
+      rfl
+    · simp only [one_mul, map_sub, coeff_mk, coeff_one, hn, ↓reduceIte]
+      rw [show n = n - 1 + 1 by exact (Nat.succ_pred hn).symm, sub_mul]
+      simp only [one_mul, map_sub, coeff_mk, Pi.one_apply, coeff_succ_X_mul, sub_self]
+
+/--
+Note that `mk 1` is the constant function `1` so the power series `1 + X + X^2 + ...`. This theorem
+states that for any `d : ℕ`, `(1 + X + X^2 + ... : S⟦X⟧) ^ (d + 1)` is equal to the power series
+`mk fun n => Nat.choose (d + n) d : S⟦X⟧`.
+-/
+theorem mk_one_pow_eq_mk_choose_add :
+    (mk 1 : S⟦X⟧) ^ (d + 1) = (mk fun n => Nat.choose (d + n) d : S⟦X⟧) := by
+  induction d with
+  | zero => ext; simp
+  | succ d hd =>
+      ext n; rw [Nat.succ_eq_add_one, pow_add, hd, pow_one, mul_comm, coeff_mul]
+      simp_rw [coeff_mk, Pi.one_apply, one_mul]; norm_cast
+      rw [Finset.sum_antidiagonal_choose_add, ← Nat.choose_succ_succ]; congr
+      exact add_right_comm d n 1
+
+/--
+The power series `mk fun n => Nat.choose (d + n) d`, whose multiplicative inverse is
+`(1 - X) ^ (d + 1)`.
+-/
+noncomputable def invOneSubPow : S⟦X⟧ˣ where
+  val := mk fun n => Nat.choose (d + n) d
+  inv := (1 - X) ^ (d + 1)
+  val_inv := by
+    rw [← mk_one_pow_eq_mk_choose_add, ← mul_pow, mk_one_mul_one_sub_eq_one]
+    exact one_pow (d + 1)
+  inv_val := by
+    rw [← mk_one_pow_eq_mk_choose_add, ← mul_pow, mul_comm, mk_one_mul_one_sub_eq_one]
+    exact one_pow (d + 1)
+
+theorem invOneSubPow_val_eq_mk_choose_add :
+    (invOneSubPow d).val = (mk fun n => Nat.choose (d + n) d : S⟦X⟧) := rfl
+
+theorem invOneSubPow_val_zero_eq_invUnitSub_one :
+    (invOneSubPow 0).val = invUnitsSub (1 : Sˣ) := by
+  rw [invOneSubPow, invUnitsSub]
+  simp only [zero_add, Nat.choose_zero_right, Nat.cast_one, one_pow, divp_one]
+
+theorem invOneSubPow_eq_inv_one_sub_pow :
+    invOneSubPow d = ⟨1 - X, (mk 1 : S⟦X⟧), by rw [mul_comm]; exact mk_one_mul_one_sub_eq_one,
```

Comment:
The statement of this theorem could be cleaned up a bit using `Units.mkOfMulEqOne`